### PR TITLE
Display the work item detail as a panel that slides out over the list view

### PR DESCRIPTION
**Pull request by nimishamukherjee**
DO NOT MERGE 
Display the work item detail as a panel that slides out over the list view (#208)

### CODE REVIEW

**Comment by bdellasc** [comment]
@sanbornsen 

**Temporary icons for list view work item types**

Josh asked me to provide you with a set of temporary set of Font Awesome icons for the different work item types. We can change them out later when we put some more thought into them:

- Value Prop = fa-gift

- Scenario = fa-certificate

- Experience = fa-map

- Fundamental = fa-bank

- Feature = fa-mouse-pointer

Here is the cheatsheet for Font Awesome that shows the names of the icons and the unicode for each icon. [http://fontawesome.io/cheatsheet/](http://fontawesome.io/cheatsheet/). Here is the PatternFly icon cheat sheet [https://www.patternfly.org/styles/icons/cheatsheet/#_](https://www.patternfly.org/styles/icons/cheatsheet/#_)

**User Icon in list view**
Josh also asked me to clarify the assigned user icons in the list. Where we show the user icon in the list, there are 3 potential states that would show here:

1. No user assigned = pficon-user with 1px stroke bounding box, icon and stroke color: #ececec (which is a lighter gray)
2. User assigned with default user icon (no custom avatar set by the user = pficon-user with no bounding box, icon color: #363636 (which is a darker gray) 
3. User assigned with custom icon = tbd, we haven't designed this interaction yet. **NOTE** In the visual design mockups, the various patterns you see in the user column just represent custom avatars that the user could set. I just put them in the mockup to give us a sense of how it will look with custom images mixed in with the default user icons. 

Hope this helps.

Thanks,
Brian

**Comment by aslakknutsen** [comment]
* Can't close slide by pressing 'Esc' on keyboard
* in consistent order of 'ok, cancel' between Title and Description during edit 
* Description 'frame' to small to see full description
* Description 'frame' 'style/type' change if you attempt to edit, then cancel. It stays as some form of textarea

**Comment by aslakknutsen** [comment]
Title with more than 6 words starts to wrap due to the position of the Status button. The edit title icon is now in a strange position

screenshot from 2016-11-05 04-53-24

File: src/app/work-item/work-item-list/work-item-list.component.scss [hunk 1]
```diff
@@ -28,5 +29,29 @@
...
     width: 100%;
-  }
+  }  
+}
+
+.detail-panel {
+  background: #ffffff;  
+  position: fixed;
+  left: auto;
+  top: 35px;
```
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: @mindreeper2420 Yes, so I did all these crap, I'll sit with @SMahil once she is back in office on Monday and update al the CSS work. Excuse my less knowledge on CSS for now!!

**Comment by sanbornsen** [comment]
@bdellasc Thanks. I am still missing out userstory from the work item types and we have no type call scenario yet. 
About the user icon, I am sure which one you are talking about!! @joshuawilson any idea?

**Comment by mindreeper2420** [comment]
There should not be any edit icons visible by default, according to the design: https://invis.io/NU8YPHLJP. The edit icons are not to appear unless the field is hovered over or in an active edit state.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+      this.save();
+      if (this.headerEditable) this.toggleHeader(); 
+      if (this.descEditable) this.toggleDescription();      
+    }
+  }
+
+  @HostListener('window:keydown', ['$event']) 
+  onKeyEvent(event: any) {
+    event = (event || window.event);
+    if (event.keyCode == 27) {
```
Comment (aslakknutsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts]: Add a comment to indicate which key by name this represents.

File: src/app/work-item/work-item-list/work-item-list.component.scss (hunk 1 above)
Comment (SMahil) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: @mindreeper2420 I will update the scss. We will be using ems for now

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.scss [hunk 3]
```diff
@@ -0,0 +1,139 @@
+@import "./../../../../assets/stylesheets/custom";
+@import "./../../../../assets/stylesheets/mixins";
+@import "./../../../../assets/stylesheets/color-variables";
+.detail-container {
+  width: 100%;
+  float: left;
+  .note{
+    color: #cccccc;
```
Comment (mindreeper2420) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.scss]: two questions:

1. Can colors be assigned variables?
2. There is not a hex code of #cccccc in [PatternFly](http://www.patternfly.org/styles/color-palette/#_). If this did not come from the visual designs, then it should be changed to match the library.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.scss [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  .detail-desc-action {
+    margin-top: $margin10;
+  }
+  .detail-action-btn {
+    font-size: $font14;
+    cursor: pointer;
+    min-width:2em;
+    text-align: $textCenter;
+    padding: $pad5;
+    background-color: #ebebeb;
```
Comment (mindreeper2420) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.scss]: same comment as above - also, some colors have variables, while others do not.

File: src/app/work-item/work-item-list/work-item-list-entry/work-item-list-entry.component.scss [hunk 5]
```diff
@@ -5,9 +5,10 @@
   cursor: pointer;
   .workItemList_workItemType {
     span{
-      font-size: $font12;
+      font-size: $font14;
```
Comment (mindreeper2420) [inline src/app/work-item/work-item-list/work-item-list-entry/work-item-list-entry.component.scss]: check with @bdellasc before changing font sizes, as the UXD Visuals are based off of particular font sizes. If the size is too small, then he should know.

File: src/app/work-item/work-item-list/work-item-list.component.scss [hunk 6]
```diff
@@ -28,5 +29,28 @@
...
+.detail-panel {
+  background: $color-pf-white;  
+  position: fixed;
+  left: auto;
+  top: 34px;
+  right: 0;
+  bottom: 0;
+  width: 50%;
+  min-width: 250px;
+  z-index: 9999;
```
Comment (mindreeper2420) [inline src/app/work-item/work-item-list/work-item-list.component.scss]: curious as to why the z-index is set so high. Are there 9998 other layers that this needs to be above?

File: src/assets/stylesheets/custom.scss [hunk 7]
```diff
@@ -20,9 +21,12 @@ $font11: 11px;
 $font12: 12px;
 $font13: 13px;
 $font14: 14px;
+$font15: 15px;
 $font16: 16px;
 $font18: 18px;
 $font20: 20px;
```
Comment (mindreeper2420) [inline src/assets/stylesheets/custom.scss]: curiosity on all of the font variables there are here:

- PatternFly has a default set of variables related to the 12px default size. They also have a set of variables that are relationship based, so that this method of multiple variables for increasing font sizes by 1px is not needed.
- Ideally, if custom variables are required for all of these fonts outside of the scope of PatternFly, then they would be increased using em or rem, not pixels.

PatternFly custom variables: https://github.com/patternfly/patternfly/blob/master/src/less/variables.less

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    <fieldset [attr.disabled]="!loggedIn ? true : null">
+      <div class="width100 pull-left mobile-bar paddingL20 paddingR20">
+        <div class="pull-left dropdown-kebab-pf detail-type-dropdown" dropdown>
+          <span dropdown-open class="wi-type-icon" almIcon [iconType]="workItem.type"></span>
+          <ul class="dropdown-menu mobMarginL20" aria-labelledby="dropdownKebabRight">
+            <li *ngFor="let type of workItemTypes" (click)="onChangeType(type.name)">
+              <a>
+                <span class="wi-type-icon dropdown-icon" almIcon [iconType]="type.name"></span>
+                <span class="dropdown-text">
+                  {{type.name.substring(7,type.length)}}
```
Comment (joshuawilson) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: @sanbornsen you may want to consider pipes here.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.scss (hunk 3 above)
Comment (joshuawilson) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.scss]: @SMahil can you please look into these questions that @mindreeper2420 has raised.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts (hunk 2 above)
Comment (joshuawilson) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.ts]: @sanbornsen please add the comment that Aslak is asking for.

File: src/app/work-item/work-item-list/work-item-list-entry/work-item-list-entry.component.scss (hunk 5 above)
Comment (joshuawilson) [inline src/app/work-item/work-item-list/work-item-list-entry/work-item-list-entry.component.scss]: @nimishamukherjee @sanbornsen This change is not related to this PR. Please remove it.

File: src/app/work-item/work-item-list/work-item-list.component.html [hunk 9]
```diff
@@ -1,12 +1,13 @@
...
 	<div class="width100 fl">
 		<div class="list-group list-view-pf width100 fl">
 			<alm-work-item-list-entry *ngFor="let workItem of workItems; let counter = index"
 				id="{{'workItemList_OuterWrap_'+counter}}"
 				class="work-item-list-entry"
-				[workItem]="workItem" 
-				(selectEvent)="onSelect($event)" 
-				(detailEvent)="onDetail($event)" 
+				[workItem]="workItem"
+        (selectEvent)="onSelect($event)"
```
Comment (joshuawilson) [inline src/app/work-item/work-item-list/work-item-list.component.html]: Why is this even a change?

File: src/app/work-item/work-item-list/work-item-list.module.ts [hunk 10]
```diff
@@ -1,21 +1,26 @@
...
+import { WorkItemDetailModule } from './work-item-detail/work-item-detail.module';
 import { WorkItemQuickAddModule } from './../work-item-quick-add/work-item-quick-add.module';
 import { WorkItemListEntryComponent } from './work-item-list-entry/work-item-list-entry.component';
 
+import { WorkItemListRoutingModule } from './work-item-list-routing.module';
+
 @NgModule({
   imports:      [
     CommonModule, 
+    WorkItemDetailModule,
```
Comment (joshuawilson) [inline src/app/work-item/work-item-list/work-item-list.module.ts]: please make these alphabetical

File: src/app/work-item/work-item.service.ts [hunk 11]
```diff
@@ -29,13 +29,13 @@ export class WorkItemService {
 
   getWorkItems(): Promise<WorkItem[]> {
     return this.http
-      .get(this.workItemUrl, {headers: this.headers})
-      .toPromise()
-      .then(response => process.env.ENV != 'inmemory' ? response.json() as WorkItem[] : response.json().data as WorkItem[])
-      .catch(this.handleError);
+    .get(this.workItemUrl, {headers: this.headers})
```
Comment (joshuawilson) [inline src/app/work-item/work-item.service.ts]: I think these should still be indented. They are coming off of "this.http".

File: src/assets/stylesheets/custom.scss (hunk 7 above)
Comment (joshuawilson) [inline src/assets/stylesheets/custom.scss]: @SMahil please review.

File: src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html (hunk 8 above)
Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: @joshuawilson Yes I'd, after the demo.

Comment (sanbornsen) [inline src/app/work-item/work-item-list/work-item-detail/work-item-detail.component.html]: @joshuawilson Actually once we have JSON api data, we'll not need to do the substring anymore. So, let's keep it like this for now.
